Docs/trl docs revisited (#3636)
recognai/rubrix: docs/_source/guides/llms/practical_guides/fine_tune.md

# Fine-tuning a Feedback Dataset

> After [collecting the responses](/guides/llms/practical_guides/collect_responses.html) from our `FeedbackDataset` we can start fine-tuning our LLMs and other models. Due to the customizability of the task, this might require setting up a custom post-processing workflow but we will provide some good toy examples for the [LLM approaches](/guides/llms/conceptual_guides/rlhf.html): pre-training, supervised fine-tuning, and reinforcement learning through human feedback (RLHF). However, we also still provide for other NLP tasks like text classification.

Reviewer: ```suggestion
After [collecting the responses](/guides/llms/practical_guides/collect_responses.html) from our `FeedbackDataset`, we can start fine-tuning our LLMs and other models. Due to the customizability of the task, this might require setting up a custom post-processing workflow, but we will provide some good toy examples for the [LLM approaches](/guides/llms/conceptual_guides/rlhf.html): pre-training, supervised fine-tuning, and reinforcement learning through human feedback (RLHF). However, we also still provide for other NLP tasks like text classification.
```

## The `ArgillaTrainer`

The `ArgillaTrainer` is a wrapper around many of our favorite NLP libraries. It provides a very intuitive abstract representation to facilitate simple training workflows using decent default pre-set configurations without having to worry about any data transformations from Argilla.

> Using the `ArgillaTrainer` is straighforward but does slightly differ per task.

Reviewer: ```suggestion
Using the `ArgillaTrainer` is straightforward, but it does differ slightly per task.
```

Reviewer 2: Maybe?

```suggestion
Using the `ArgillaTrainer` is straightforward, but it slightly differs per task.
```

> 1. First, we then define a `TaskMapping`. This is done using a custom `formatting_func`. However, tasks like Text Classification can also be defined using default definitions using the FeedbackDataset fields and questions. These mappings are then used for retrieving data from a dataset and initializing the training. We also offer some ideas for [unifying data](/guides/llms/practical_guides/collect_responses) out of the box.

Reviewer: ```suggestion
1. First, we define a `TrainingTask`. This is done using a custom `formatting_func`. However, tasks like Text Classification can also be defined using default definitions using the `FeedbackDataset` fields and questions. These tasks are then used for retrieving data from a dataset and initializing the training. We also offer some ideas for [unifying data](/guides/llms/practical_guides/collect_responses) out of the box.
```

> 2. Next, we initialize the `ArgillaTrainer` and forward the task mapping and training framework. Internally, this uses the `FeedbackData.prepare_for_training`-method to format the data according to the expectations from the framework. Some other interesting methods are:

Reviewer: ```suggestion
2. Next, we initialize the `ArgillaTrainer` and forward the task and training framework. Internally, this uses the `FeedbackData.prepare_for_training`-method to format the data according to the expectations from the framework. Some other interesting methods are:
```

>    1. `ArgillaTrainer.update_config` to change framework specifc training paramaeters.

Reviewer: ```suggestion
   1. `ArgillaTrainer.update_config` to change framework specific training parameters.
```

   2. `ArgillaTrainer.train` to start training.
   3. `ArgillTrainer.predict` to run inference.

> Underneath, you can see the happy flow for using the `ArgillaTrainer`.

Reviewer: Happy flow?

Author: Yes, the easiest way something is likely to work.

Reviewer: Haha awesome

    repo_id="argilla/emotion"
)
task = TrainingTask.for_text_classification(
    text=dataset.field_by_name("text"),
    label=dataset.question_by_name("label"),
)
trainer = ArgillaTrainer(
    dataset=dataset,
    task=task,
    framework="setfit"
)
trainer.update_config(num_iterations=1)
trainer.train(output_dir="my_setfit_model")
trainer.predict("This is awesome!")
```

### Supported Frameworks

We plan on adding more support for other tasks and frameworks so feel free to reach out on our Slack or GitHub to help us prioritize each task.

| Task/Framework                   | TRL  | OpenAI | AutoTrain | SetFit | spaCy | Transformers | PEFT |
|:--------------------------------|:-----|:-------|:----------|:-------|:------|:-------------|:-----|
| Text Classification             |      | ✔️      | ✔️         | ✔️      | ✔️     | ✔️            | ✔️    |
| Supervised Fine-tuning          | ✔️    |        |           |        |       |              |      |
| Reward Modeling                 | ✔️    |        |           |        |       |              |      |
| Proximal Policy Optimization    | ✔️    |        |           |        |       |              |      |

> | Direct Performance Optimization | ✔️    |        |           |        |       |              |      |

Reviewer: ```suggestion
| Direct Preference Optimization  | ✔️    |        |           |        |       |              |      |
```

```{note}

> We also offer support for  Token Classification using our `TokenClassifcationDataset` but this is shown in [a section](/guides/train_a_model) about our older dataset-types.

Reviewer: ```suggestion
We also offer support for Token Classification using our `TokenClassifcationDataset` but this is shown in [a section](/guides/train_a_model) about our older dataset-types.
```

```

#### Training Configs

The trainer also has an `ArgillaTrainer.update_config()` method, which maps a dict with `**kwargs` to the respective framework. So, these can be derived from the underlying framework that was used to initialize the trainer. Underneath, you can find an overview of these variables for the supported frameworks.

```{note}
Note that you don't need to pass all of them directly and that the values below are their default configurations.
```

```{include} /_common/tabs/train_update_config.md
```

> ### The `TaskMapping`

Reviewer: ```suggestion
### The `TrainingTask`
```

> A `TaskMapping` is used to define how the data should be processed and formatted according to the associated task and framework. Each task has its own `TaskMapping.for_*`-classmethod and the data formatting can always be defined using a custom `formatting_func`. However, simpler tasks like Text Classification can also be defined using default definitions. These directly use the fields and questions from the FeedbackDataset configuration to infer how to prepare the data. Underneath you can find an overview of the `TaskMapping`` requirements.

Reviewer: ```suggestion
A `TrainingTask` is used to define how the data should be processed and formatted according to the associated task and framework. Each task has its own `TrainingTask.for_*`-classmethod and the data formatting can always be defined using a custom `formatting_func`. However, simpler tasks like Text Classification can also be defined using default definitions. These directly use the fields and questions from the FeedbackDataset configuration to infer how to prepare the data. Underneath you can find an overview of the `TrainingTask` requirements.
```

> | Method                             | Content          | Return formatting_func                                     | Default           |

Reviewer: Maybe "Default" should be e.g. "Has a default"?

Author: I considered that, but it slightly messes up the formatting.

Reviewer: Makes sense

> | Method                             | Content          | Return formatting_func                                     | Default           |

Reviewer: ```suggestion
| Method                             | Content          | `formatting_func` return type                                     | Default           |
```

> |:-----------------------------------|:-----------------|:-----------------------------------------------------------|:------------------|
> | for_text_classification            | `text-label`     | `Union[Tuple[str, str], Tuple[str, List[str]]]`            | ✔️                 |
> | for_supervised_fine_tuning         | `text`           | `str`                                                      | ✗                 |
> | for_reward_modelling               | `chosen-rejected`| `Tuple[str, str]`                                          | ✗                 |
> | for_proximal_policy_optimization   | `text`           | `Union[str, Iterator[str]]`                                | ✗                 |
> | for_direct_performance_optimization| `prompt-chosen-rejected`                 | `Tuple[str, str]`                                          | ✗                 |

Reviewer: ```suggestion
|:-----------------------------------|:-----------------|:-----------------------------------------------------------|:------------------|
| for_text_classification            | `text-label`     | `Union[Tuple[str, str], Tuple[str, List[str]]]`            | ✔️                 |
| for_supervised_fine_tuning         | `text`           | `Optional[Union[str, Iterator[str]]]`                                                      | ✗                 |
| for_reward_modelling               | `chosen-rejected`| `Optional[Union[Tuple[str, str], Iterator[Tuple[str, str]]]]`                                          | ✗                 |
| for_proximal_policy_optimization   | `text`           | `Optional[Union[str, Iterator[str]]]`                                | ✗                 |
| for_direct_performance_optimization| `prompt-chosen-rejected`                 | `Optional[Union[Tuple[str, str, str], Iterator[Tuple[str, str, str]]]]`                                          | ✗                 |
```

They all accept iterators and `None`. Not sure about text classification though.

```

:::

:::{tab-item} Multi-label
Multi-label text classification is generally more complex than single-label classification due to the challenge of determining and predicting multiple relevant labels for each text. It finds applications in various domains, including document tagging, topic labeling, and content recommendation systems. For example, in customer care, a multi-label text classification task would involve assigning topics such as "new_card," "mortgage," or "opening_hours" to texts based on their content.

```{tip}
For a multi-label scenario it is recommended to add some examples without any labels to improve model performance.
```

```batch
"The help for my application of a new card and mortgage was great", ["new_card", "mortgage"]
```

:::

::::

#### Training

Text classification is one of the most widely supported training tasks tasks within NLP. For example purposes we will use our [Stackoverflow demo dataset](https://huggingface.co/datasets/argilla/stackoverflow_feedback_demo).

**Data Preparation**

```python
from argilla.feedback import FeedbackDataset

dataset = FeedbackDataset.from_huggingface(

>     repo_id="argilla/emotion"

Reviewer: ```suggestion
    repo_id="argilla/stackoverflow_feedback_demo"
```
Then we should use the Stack overflow demo right? Or is the text above wrong?

Reviewer: Or the other way around, that we say we'll use the emotion dataset for the demo.

```{note}
An overview of the unifcation measures can be found [here](/guides/llms/practical_guides/collect_responses). The `RatingQuestion` and `RankingQuestion` can be unified using a "majority"-, "min"-, "max"- or "disagreement"-strategy. Both the `LabelQuestion` and `MultiLabelQuestion` can be resolved using a "majority"-, or "disagreement"-strategy.
```

```python
from argilla.feedback import FeedbackDataset, TrainingTask

dataset = FeedbackDataset.from_huggingface(
    repo_id="argilla/emotion"
)
task = TrainingTask.for_text_classification(
    text=dataset.field_by_name("text"),
    label=dataset.question_by_name("label"),
    label_strategy=None # defaults presets
)
```

:::

:::{tab-item} formatting_func
We offer the option to provide a `formatting_func` to the `TrainingTask.for_text_classification`. This function is applied to each sample in the dataset and can be used for more advanced preprocessing and data formatting. The function should return a tuple of `(text, label)`  as `Tuple[str, str]` or `Tuple[str, List[str]]`.

```python
from argilla.feedback import FeedbackDataset, TrainingTask

> dataset = FeedbackDataset.from_huggingface(
>     repo_id="argilla/emotion"
> )

Reviewer: Perhaps these should be on one line?

```suggestion
dataset = FeedbackDataset.from_huggingface(repo_id="argilla/emotion")
```

Author: https://huggingface.co/datasets/argilla/emotion

def formatting_func(sample):
    text = sample["text"]

>     values = [resp["value"] for resp in sample["label"]]

Reviewer: ```suggestion
    # Choose the most common label
    values = [resp["value"] for resp in sample["label"]]
```
Maybe something like this to clarify what this does?

    counter = Counter(values)
    if counter:
        most_common = counter.most_common()
        max_frequency = most_common[0][1]
        most_common_elements = [
            element for element, frequency in most_common if frequency == max_frequency
        ]
        label = random.choice(most_common_elements)
        return (text, label)
    else:
        return None

> task = TrainingTask.for_text_classification(
>     formatting_func=formatting_func,
> )

Reviewer: Should this be just on one line?
```suggestion
task = TrainingTask.for_text_classification(formatting_func=formatting_func)
```

```

:::

::::

We can then define our `ArgillaTrainer` for any of [the supported frameworks](fine_tune.md#training-configs) and [customize the training config](#supported-frameworks) using `ArgillaTrainer.update_config`.

```python
from argilla.feedback import ArgillaTrainer

trainer = ArgillaTrainer(
    dataset=feedback_dataset,
    task=task,
    framework="spacy",

>     fetch_records=False,

Reviewer: ```suggestion
```
We removed this parameter

```

### Pre-training

#### Background

When talking about pre-training, we generally talk about a simple `prompt-completion` task, where we need the model to pick up on basic statistics of the language it is learning. Given that you are familiar with Spanish cuisine and the prompt sentence, `The base ingredient of paella is ___`, you know that the word in the `___` is much more likely to be `rice` than `apples`.  So, you are basically training a causal language model or text generation model.

```{note}
This is an unsupervised approach hence we only infer training data from a basic sentence like `The base ingredient of paella is rice.` by starting with the word `The`, and from there unwrapping the sentence step by step.
```

#### Training

Many training datasets for this task can be found online (e.g., [Hugging Face](https://huggingface.co/datasets?task_categories=task_categories:text-generation&sort=downloads)). You can either upload this in the right Argilla format but it might be needed to collect and fine-tune additional data with Argilla. So we, therefore, provide a basic setup underneath which should help you to start gathering or preparing pre-training data.

```{note}

> When it comes to pre-training an LLM, we generally do not need data of highest quality, but it is always smart to use domain-specfic data and to avoid data that might lead to undecired effect like hallucination and bias.

Reviewer: ```suggestion
When it comes to pre-training an LLM, we generally do not need data of highest quality, but it is always smart to use domain-specfic data and to avoid data that might lead to undesired effects like hallucination and bias.
```

```

First, create a `FeedbackDataset` with records.

```python
import argilla as rg

> # create promp-completion dataset

Reviewer: ```suggestion
# create prompt-completion dataset
```

dataset = rg.FeedbackDataset(
    guidelines="Please, complete the following prompt fields with a brief text answer.",
    fields=[
        rg.TextField(name="prompt"),
    ],
    questions=[
        rg.TextQuestion(name="completion", title="Add a brief text answer."),
    ]
)

# create a Feedback Records
record = rg.FeedbackRecord(
    fields={
        "prompt": "The base ingredient of paella is rice."
    }
)

dataset.add_records([record])
```

Then push it to Argilla via `push_to_argilla`.

::::{tab-set}

:::{tab-item} Argilla 1.14.0 or higher
```python
remote_dataset = dataset.push_to_argilla(name="pre-training")
```
:::

:::{tab-item} Lower than Argilla 1.14.0
```python
dataset.push_to_argilla(name="pre-training")
```
:::
::::

And, finally, load the `FeedbackDataset` from Argilla.

```python
import argilla as rg
from datasets import Dataset

dataset = rg.FeedbackDataset.from_argilla("pre-training")
prompts = {"prompt": [record.fields.get("prompt") for record in dataset.records]}
dataset = Dataset.from_dict(prompts)
dataset
# Dataset({
#     features: ['prompt'],
#     num_rows: 1
# })
```

> There are many ways and great packages to deal with this `pre-training` phase, but generally, NLP training frameworks like [KerasNLP](https://keras.io/keras_nlp/) and [Hugging Face](https://huggingface.co/) offer great out-of-the-box methods for training a causal language model. In our guide, we will refer to the great docs off using Hugging Face `transformers` and `datasets` library and prepare our training data in the format they require for [training a causal language model](https://huggingface.co/learn/nlp-course/chapter7/6#training-a-causal-language-model-from-scratch).

Reviewer: ```suggestion
There are many ways and great packages to deal with this `pre-training` phase, but generally, NLP training frameworks like [KerasNLP](https://keras.io/keras_nlp/) and [Hugging Face](https://huggingface.co/) offer great out-of-the-box methods for training a causal language model. In our guide, we will refer to the great docs of the Hugging Face `transformers` and `datasets` libraries and prepare our training data in the format they require for [training a causal language model](https://huggingface.co/learn/nlp-course/chapter7/6#training-a-causal-language-model-from-scratch).
```

Ultimately, the choice between these two approaches depends on the specific requirements of the application and the desired level of control over the model's output. By employing the appropriate fine-tuning strategy, we can enhance the model's performance and make it more suitable for a wide range of applications and use cases.

#### Training

There are many good libraries to help with this step, however, we are a fan of the [Transformer Reinforcement Learning (TRL)](https://huggingface.co/docs/trl) package, [Transformer Reinforcement Learning X (TRLX)](https://github.com/CarperAI/trlx),and the no-code [Hugging Face AutoTrain](https://huggingface.co/spaces/autotrain-projects/autotrain-advanced) for fine-tuning. In both cases, we need a backbone model, obtained from the [pre-training step](#pre-training) and for example purposes we will use our [curated Dolly dataset](https://huggingface.co/datasets/argilla/databricks-dolly-15k-curated-en).

```{note}

> This dataset only contains a single annotator response per record. We gave some sugggestions on dealing with [responses from multiple annotators](/guides/llms/practical_guides/collect_responses).

Reviewer: ```suggestion
This dataset only contains a single annotator response per record. We gave some suggestions on dealing with [responses from multiple annotators](/guides/llms/practical_guides/collect_responses).
```

```

::::{tab-set}

:::{tab-item} TRL

The [Transformer Reinforcement Learning (TRL)](https://huggingface.co/docs/trl) package provides a flexible and customizable framework for fine-tuning models. It allows users to have fine-grained control over the training process, enabling them to define their functions and to further specify the desired behavior of the model. This approach requires a deeper understanding of reinforcement learning concepts and techniques, as well as more careful experimentation. It is best suited for users who have experience in reinforcement learning and want fine-grained control over the training process. Additionally, it directly integrates with [Parameter-Efficient Fine-Tuning](https://huggingface.co/docs/peft/index) (PEFT) decreasing the computational complexity of this step of training an LLM.

**Data Preparation**

```python
import argilla as rg
from datasets import Dataset

feedback_dataset = rg.FeedbackDataset.from_huggingface("argilla/databricks-dolly-15k-curated-en")
```

> We offer the option to provide a `formatting_func` to the `TrainingTask.for_supervised_fine_tuning`. This function is applied to each sample in the dataset and can be used for anced preprocessing and data formatting. The function should return a tuple of `text`  as `str`.

Reviewer: ```suggestion
We offer the option to provide a `formatting_func` to the `TrainingTask.for_supervised_fine_tuning`. This function is applied to each sample in the dataset and can be used for anced preprocessing and data formatting. The function should return a tuple of `text` as `str`.
```

Reviewer: Also, what is `anced preprocessing`?

trainer = trlx.train('gpt2', samples=samples)
```

:::

::::

### Reward Modeling

#### Background

A Reward Model (RM) is used to rate responses in alignment with human preferences and afterwards using this RM to fine-tune the LLM with the associated scores. Fine-tuning using a Reward Model can be done in different ways. We can either get the annotator to rate output completely manually, we can use a simple heuristic or we can use a stochastic preference model. Both [TRL](https://huggingface.co/docs/trl) and [TRLX](https://github.com/CarperAI/trlx) provide decent options for incorporating rewards. The [DeepSpeed library of Microsoft](https://github.com/microsoft/DeepSpeed/tree/master/blogs/deepspeed-chat) is a worthy mention too but will not be covered in our docs.

The data required for these steps need to be used as comparison data to showcase the preference for the generated prompts. A good example is our [curated Dolly dataset](https://huggingface.co/datasets/argilla/databricks-dolly-15k-curated-en), where we assumed that updated responses get preference over the older ones.

```{note}

> The Dolly original dataset contained a lot of reference indicators "[]", which enforces the model to hallicinate and create references.

Reviewer: ```suggestion
The Dolly original dataset contained a lot of reference indicators such as "[1]", which causes the model to hallucinate and incorrectly create references.
```

### Context
Virgin Australia, the trading name of Virgin Australia Airlines Pty Ltd, is an Australian-based airline.
It is the largest airline by fleet size to use the Virgin brand.
It commenced services on 31 August 2000 as Virgin Blue, with two aircraft on a single route.
It suddenly found itself as a major airline in Australia's domestic market after the collapse of Ansett Australia in September 2001.
The airline has since grown to directly serve 32 cities in Australia, from hubs in Brisbane, Melbourne and Sydney.

### Response:
Virgin Australia commenced services on 31 August 2000 as Virgin Blue, with two aircraft on a single route.
```

:::

::::

#### Training

The data required for these steps need to be used as comparison data to showcase the preference for the generated prompts. Like before, we will use our [curated Dolly dataset](https://huggingface.co/datasets/argilla/databricks-dolly-15k-curated-en), where we assumed that updated responses get preference over the older ones.

```python
import argilla as rg

feedback_dataset = rg.FeedbackDataset.from_huggingface("argilla/databricks-dolly-15k-curated-en")
```

::::{tab-set}

:::{tab-item} TRL

> [TRL](https://huggingface.co/docs/trl) implements reward modeling, which can be used via the `ArgillaTrainer` class. We offer the option to provide a `formatting_func` to the `TrainingTask.for_reward_modelling`. This function is applied to each sample in the dataset and can be used for preprocessing and data formatting. The function should return a tuple of `chosen-rejected`-pairs  as `Tuple[str, str]`. To determine which response from the FeedbackDataset is superior, we can use the user annotations.

Reviewer: ```suggestion
[TRL](https://huggingface.co/docs/trl) implements reward modeling, which can be used via the `ArgillaTrainer` class. We offer the option to provide a `formatting_func` to the `TrainingTask.for_reward_modeling`. This function is applied to each sample in the dataset and can be used for preprocessing and data formatting. The function should return a tuple of `chosen-rejected`-pairs  as `Tuple[str, str]`. To determine which response from the FeedbackDataset is superior, we can use the user annotations.
```

```{note}

> The formatting function can also return `None` or a list of tuples. The `None` may be used if the annotations indicate that the text is low quality or harmful, and the latter could be used if multiple annotators provide additional written responses, resulting in multiple good `chosen-rejected` pairs.

Reviewer: ```suggestion
The formatting function can also return `None` or a list of tuples. The `None` may be used if the annotations indicate that the text is low quality or harmful, and the list of tuples could be used if multiple annotators provide additional written responses, resulting in multiple good `chosen-rejected` pairs.
```

# >> 2.2948970794677734
```
As expected, the good response has a higher score than the worse response.
:::

::::

### Proximal Policy Optimization

#### Background

The [TRL](https://huggingface.co/docs/trl) library implements the last step of RLHF: Proximal Policy Optimization (PPO). It requires prompts, which are then fed through the model being finetuned. Its results are passed through a reward model. Lastly, the prompts, responses and rewards are used to update the model through reinforcement learning.

```{note}

> PPO requires a trained trainer supervised fine-tuned model and reward model to function. Take a look at that task outlines above to train your own models.

Reviewer: ```suggestion
PPO requires a trained supervised fine-tuned model and reward model to work. Take a look at that task outlines above to train your own models.
```

```

#### Training

> The data required for these steps need to be used as comparison data to showcase the preference for the generated prompts. Like before, we will use our [curated Dolly dataset](https://huggingface.co/datasets/argilla/databricks-dolly-15k-curated-en), where we assumed that updated responses get preference over the older ones.

Reviewer: ```suggestion
The data required for these steps needs to be used as comparison data to showcase the preference for the generated prompts. Like before, we will use our [curated Dolly dataset](https://huggingface.co/datasets/argilla/databricks-dolly-15k-curated-en), where we assumed that updated responses get preference over the older ones.
```

outputs = model.generate(**encoding, max_new_tokens=30)
output_text = tokenizer.decode(outputs[0])
print(output_text)
# Yes it is, toads are a sub-classification of frogs.
```

### Direct Preference Optimization

#### Background

The [TRL](https://huggingface.co/docs/trl) library implements and alternative way to incorporate human feedback into an LLM which is called Direct Preference Optimization (DPO). This approach skips the step of training a separate reward model and directly uses the preference data during training as measure for optimization of human feedback. In order to properly use th

```{note}
DPO requires a trained supervised fine-tuned model to function. Take a look at that task outline above to train your own model.
```

```{include} /_common/dolly_dataset_info.md
```

> In case of training an PPO, we then use the prompt and context data and correct the generated response from the SFT model by using the reward model. Hence, we will need to format the following `text`.

Reviewer: ```suggestion
In case of training using PPO, we then use the prompt and context data and correct the generated response from the SFT model by using the reward model. Hence, we will need to format the following `text`.
```

We will start with our a basic example of a formatting function. For DPO it should return `prompt-chosen-rejected`-pairs, where the prompt is formatted according to a template.

```python
from argilla.feedback import TrainingTask
from typing import Dict, Any, Iterator

template = """\
### Instruction: {instruction}\n
### Context: {context}\n
### Response: {response}"""

def formatting_func(sample: Dict[str, Any]) -> Iterator[Tuple[str, str]]:
    # Our annotators were asked to provide new responses, which we assume are better than the originals
    og_instruction = sample["original-instruction"]
    og_context = sample["original-context"]
    rejected = sample["original-response"]
    prompt = template.format(instruction=og_instruction, context=og_context, response="")

    for instruction, context, response in zip(sample["new-instruction"], sample["new-context"], sample["new-response"]):
        if response["status"] == "submitted":
            chosen = response["value"]
            if chosen != rejected:
                yield prompt, chosen, rejected

task = TrainingTask.for_direct_preference_optimization(formatting_func=formatting_func)
```

**ArgillaTrainer**

> Instead of using this dataset, we'll use the task directly with our `FeedbackDataset` in the `ArgillaTrainer`. In contrary to PPO, we do not need to specify any reward model, because this preference modeling is inferred internally by the DPO-algorithm.

Reviewer: ```suggestion
We'll use the task directly with our `FeedbackDataset` in the `ArgillaTrainer`. In contrary to PPO, we do not need to specify any reward model, because this preference modeling is inferred internally by the DPO-algorithm.
```
This was taken from somewhere that followed a snippet of how to observe this dataset. Without that, we don't need this "intro"

outputs = model.generate(**encoding, max_new_tokens=30)
output_text = tokenizer.decode(outputs[0])
print(output_text)

> # Yes it is because toads are a sub-classification of frogs.

Reviewer: ```suggestion
# Yes it is, toads are a sub-classification of frogs.
```
